### PR TITLE
`px` to rzshell

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

3/n of https://github.com/rizinorg/rizin/pull/2813

- `pxd` -> `pxdw`
- `pxd1` -> `pxd`
- `pxd2` -> `pxdh`
- `pxd4` -> `pxdw`
- `pxd8` -> `pxdq`

- `pxAv` -> `pxAl`
- `pxt.` -> removed ?
- `px/` -> keep
- `pxf` -> TODO: not found old code ?

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: librz/core/cmd/cmd_print.c [hunk 1]
```diff
@@ -4493,90 +4455,85 @@ static char *__op_refs(RzCore *core, RzAnalysisOp *op, int n) {
...
 }
 
-static void cmd_pxr(RzCore *core, int len, int mode, int wordsize, const char *arg) {
-	PJ *pj = NULL;
-	RzTable *t = NULL;
-	if (mode == ',') {
-		t = rz_table_new();
+static inline char *__refs(RzCore *core, ut64 x) {
+	char *refs = NULL;
+	if (core->print->hasrefs) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c]: You could reverse the condition to make it simpler

File: librz/core/cmd/cmd_print.c [hunk 2]
```diff
@@ -4493,90 +4455,85 @@ static char *__op_refs(RzCore *core, RzAnalysisOp *op, int n) {
...
-		if (t) {
-			rz_table_query(t, arg ? arg + 1 : NULL);
-			char *s = rz_table_tostring(t);
-			rz_cons_println(s);
-			free(s);
-			rz_table_free(t);
+		pj_end(pj);
+	} else if (mode == RZ_OUTPUT_MODE_QUIET) {
+		for (ut64 i = 0; i + wordsize < end; i += wordsize) {
+			//			ut64 addr = core->offset + i;
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c]: Remove this maybe?

File: librz/core/cmd/cmd_print.c [hunk 3]
```diff
@@ -5575,509 +5540,12 @@ RZ_IPI int rz_cmd_print(void *data, const char *input) {
 	}
 		rz_cons_break_push(NULL, NULL);
 		switch (input[1]) {
-		case 'j': // "pxj"
-			rz_core_print_jsondump(core, core->block, core->blocksize, 8);
-			break;
 		case '/': // "px/"
 			rz_core_print_examine(core, input + 2);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:5544]: This wasn't converted, it seems?

File: librz/core/cmd/cmd_print.c [hunk 4]
```diff
@@ -6199,7 +5667,7 @@ RZ_IPI int rz_cmd_print(void *data, const char *input) {
 }
 
 RZ_IPI int rz_cmd_hexdump(void *data, const char *input) {
-	return rz_cmd_print(data, input - 1);
+	return rz_core_cmdf(data, "px%s", input);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:5671]: Please add `TODO: Use the API directly` here

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd/cmd_print.c (hunk 3 above)
Comment (b1llow) [reply librz/core/cmd/cmd_print.c:5544]: Yes, it's too flexible, so keep it as `RZ_CMD_DESC_TYPE_OLDINPUT`

```
[0x00005ef0]> px/?
Format is x/[num][format][size]
Num specifies the number of format elements to display
Format letters are o(octal), x(hex), d(decimal), u(unsigned decimal),
  t(binary), f(float), a(address), i(instruction), c(char) and s(string),
  T(OSType), A(floating point values in hex).
Size letters are b(byte), h(halfword), w(word), g(giant, 8 bytes).

```

Comment (XVilka) [reply librz/core/cmd/cmd_print.c:5544]: @ret2libc what do you think?

**Review by XVilka — APPROVED** [review]

File: librz/core/cmd/cmd_print.c (hunk 3 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:5544]: i like it, much more flexible

Comment (wargio) [reply librz/core/cmd/cmd_print.c:5544]: usually `giant` is `double-word`

Comment (ret2libc) [reply librz/core/cmd/cmd_print.c:5544]: Does it even work?

**Review by ret2libc — CHANGES_REQUESTED** [review]
Will review

File: librz/core/cprint.c [hunk 5]
```diff
@@ -158,21 +158,289 @@ RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len) {
...
+ */
+static inline bool rz_cons_print_own(char *str) {
+	if (!str) {
+		return false;
+	}
+	rz_cons_print(str);
+	free(str);
+	return true;
+}
+
```
Comment (ret2libc) [inline librz/core/cprint.c]: I think this is not really necessary.
Also, probably if the strings you pass to this function are NULL you may want to print a specific error message to the users, telling them X is wrong.

File: librz/include/rz_core.h [hunk 6]
```diff
@@ -1056,12 +1056,24 @@ RZ_API bool rz_core_meta_string_add(RzCore *core, ut64 addr, ut64 size, RzStrEnc
...
 RZ_API RZ_OWN char *rz_core_hex_of_assembly(RzCore *core, const char *assembly);
 RZ_API RZ_OWN char *rz_core_esil_of_assembly(RzCore *core, const char *assembly);
 RZ_API RZ_OWN char *rz_core_assembly_of_hex(RzCore *core, ut8 *hex, int len);
 RZ_API RZ_OWN char *rz_core_esil_of_hex(RzCore *core, ut8 *hex, int len);
 
+RZ_API RZ_OWN char *rz_core_print_hexdump_diff_str(RZ_NONNULL RzCore *core, ut64 aa, ut64 ba, ut64 len);
+RZ_API RZ_OWN char *rz_core_print_dump_str(RZ_NONNULL RzCore *core, RzOutputMode mode, ut64 addr, ut8 n, int len, RzCorePrintFormatType format);
+RZ_API RZ_OWN char *rz_core_print_hexdump_or_hexdiff_str(RZ_NONNULL RzCore *core, RzOutputMode mode, ut64 addr, int len, bool use_comment);
+RZ_API RZ_OWN char *rz_core_print_hexdump_byline_str(RZ_NONNULL RzCore *core, bool hex_offset, ut64 addr, int len, ut8 size);
+
```
Comment (ret2libc) [inline librz/include/rz_core.h:1076]: IIRC we said these functions should actually print and not be called `_str` (probably `_print`). Cutter or others can use internal functions like `rz_print_hexdiff_str` and similar, right?

Comment (XVilka) [reply librz/include/rz_core.h:1076]: @ret2libc if they print they should not be an API, if they should be an API they should return a string. Having an API returning nothing and just printing to the screen is absolutely useless from the user point of view.

File: librz/core/cmd/cmd_print.c (hunk 3 above)
Comment (b1llow) [reply librz/core/cmd/cmd_print.c:5544]:
```
[0x00005ef0]> x?
Usage: px[0afoswqWqQ][f]   # Print heXadecimal
| px                show hexdump
| px/               same as x/ in gdb (help x)
| px0               8bit hexpair list of bytes until zero byte
| pxa               show annotated hexdump
| pxA[?]            show op analysis color map
| pxb               dump bits in hexdump form
| pxc               show hexdump with comments
| pxd[?1248]        signed integer dump (1 byte, 2 and 4)
| pxe               emoji hexdump! :)
| pxf               show hexdump of current function
| pxh               show hexadecimal half-words dump (16bit)
| pxH               same as above, but one per line
| pxi               HexII compact binary representation
| pxl               display N lines (rows) of hexdump
| pxo               show octal dump
| pxq               show hexadecimal quad-words dump (64bit)
| pxQ[q]            same as above, but one per line
| pxr[1248][qj]     show hexword references (q=quiet, j=json)
| pxs               show hexadecimal in sparse mode
| pxt[*.] [origin]  show delta pointer table in rizin commands
| pxw               show hexadecimal words dump (32bit)
| pxW[q]            same as above, but one per line (q=quiet)
| pxx               show N bytes of hex-less hexdump
| pxX               show N words of hex-less hexdump
[0x00005ef0]> x/
[0x00005ef0]> x/?
Format is x/[num][format][size]
Num specifies the number of format elements to display
Format letters are o(octal), x(hex), d(decimal), u(unsigned decimal),
  t(binary), f(float), a(address), i(instruction), c(char) and s(string),
  T(OSType), A(floating point values in hex).
Size letters are b(byte), h(halfword), w(word), g(giant, 8 bytes).
[0x00005ef0]> x/10fb
0x00005ef0 = -2.05176598e+35
0x00005ef4 = -2.43060218e-33
0x00005ef8 = -2.41187068e-33
0x00005efc = -1.93742019e+22
0x00005f00 = 3397.05859
0x00005f04 = -728067.062
0x00005f08 = 5.08824453e+10
0x00005f0c = -nan
0x00005f10 = 6.96386875e-38
0x00005f14 = 5.25126609e-11
```

It looks OK

File: librz/include/rz_core.h (hunk 6 above)
Comment (b1llow) [reply librz/include/rz_core.h:1076]: I moved these to `librz/core/core_private.h`

Comment (ret2libc) [reply librz/include/rz_core.h:1076]: @XVilka we use those kind of rzcore APIs in cases where we need to print the same things in different places (e.g. `rz_core_asm_plugins_print`). But yes, if we think it's not needed right now, let's not make them APIs.

**Review by ret2libc — APPROVED** [review]

File: librz/core/cmd/cmd_print.c (hunk 3 above)
Comment (XVilka) [reply librz/core/cmd/cmd_print.c:5544]: I added entry about it to https://github.com/rizinorg/rizin/issues/1590, to not forget

**Pull request merged**